File: go/spore-eval/report/report.go
// Package report defines the comparison + recommendation types and their
// derivation (Rules 19-25).
package report

import (
	"fmt"
	"math"

	sporecore "github.com/squirrelsoft-dev/spore-core/go/spore-core"
	"github.com/squirrelsoft-dev/spore-core/go/spore-eval/metricmap"
	"github.com/squirrelsoft-dev/spore-core/go/spore-eval/stats"
)

// ComparisonDirection is whether the candidate is better/worse/unchanged on a
// metric, relative to the metric's optimization direction (Rule 22).
type ComparisonDirection string

const (
	// DirectionBetter — candidate improves the metric.
	DirectionBetter ComparisonDirection = "better"
	// DirectionWorse — candidate regresses the metric.
	DirectionWorse ComparisonDirection = "worse"
	// DirectionNoChange — within epsilon of no change.
	DirectionNoChange ComparisonDirection = "no_change"
)

// MetricComparison is one metric's baseline-vs-candidate comparison
// (Rules 19-22).
type MetricComparison struct {
	MetricName string            `json:"metric_name"`
	Baseline   stats.MetricStats `json:"baseline"`
	Candidate  stats.MetricStats `json:"candidate"`
	// Delta is candidate.mean - baseline.mean.
	Delta float64 `json:"delta"`
	// PValue is Welch's t-test two-sided p-value (Rule 20).
	PValue float64 `json:"p_value"`
	// CI is the bootstrap CI for non-deterministic metrics (Rule 21).
	CI        *stats.ConfidenceInterval `json:"ci,omitempty"`
	Direction ComparisonDirection       `json:"direction"`
}

// RecommendationKind discriminates Recommendation variants.
type RecommendationKind string

const (
	// RecAdopt — adopt the candidate; primary improved with p < 0.05.
	RecAdopt RecommendationKind = "adopt"
	// RecReject — candidate is clearly worse on the primary metric.
	RecReject RecommendationKind = "reject"
	// RecNeedsMoreRuns — not enough runs to call it (p >= 0.05).
	RecNeedsMoreRuns RecommendationKind = "needs_more_runs"
	// RecAmbiguous — mixed signals across metrics.
	RecAmbiguous RecommendationKind = "ambiguous"
)

// Recommendation is the runner's recommendation (Rules 23-24).
type Recommendation struct {
	Kind RecommendationKind `json:"kind"`
	// adopt
	ConfigID   string  `json:"-"`
	Confidence float64 `json:"-"`
	// reject
	Reason string `json:"-"`
	// needs_more_runs
	CurrentN     uint32 `json:"-"`
	RecommendedN uint32 `json:"-"`
	// ambiguous
	Tradeoffs []string `json:"-"`
}

// MarshalJSON serialises as a flat tagged object keyed by "kind".
func (r Recommendation) MarshalJSON() ([]byte, error) {
	switch r.Kind {
	case RecAdopt:
		return jsonMarshal(struct {
			Kind       RecommendationKind `json:"kind"`
			ConfigID   string             `json:"config_id"`
			Confidence float64            `json:"confidence"`
		}{r.Kind, r.ConfigID, r.Confidence})
	case RecReject:
		return jsonMarshal(struct {
			Kind   RecommendationKind `json:"kind"`
			Reason string             `json:"reason"`
		}{r.Kind, r.Reason})
	case RecNeedsMoreRuns:
		return jsonMarshal(struct {
			Kind         RecommendationKind `json:"kind"`
			CurrentN     uint32             `json:"current_n"`
			RecommendedN uint32             `json:"recommended_n"`
		}{r.Kind, r.CurrentN, r.RecommendedN})
	case RecAmbiguous:
		tr := r.Tradeoffs
		if tr == nil {
			tr = []string{}
		}
		return jsonMarshal(struct {
			Kind      RecommendationKind `json:"kind"`
			Tradeoffs []string           `json:"tradeoffs"`
		}{r.Kind, tr})
	default:
		return nil, fmt.Errorf("Recommendation: unknown kind %q", r.Kind)
	}
}

// UnmarshalJSON decodes the flat tagged form.
func (r *Recommendation) UnmarshalJSON(data []byte) error {
	var probe struct {
		Kind         RecommendationKind `json:"kind"`
		ConfigID     string             `json:"config_id"`
		Confidence   float64            `json:"confidence"`
		Reason       string             `json:"reason"`
		CurrentN     uint32             `json:"current_n"`
		RecommendedN uint32             `json:"recommended_n"`
		Tradeoffs    []string           `json:"tradeoffs"`
	}
	if err := jsonUnmarshal(data, &probe); err != nil {
		return err
	}
	r.Kind = probe.Kind
	r.ConfigID = probe.ConfigID
	r.Confidence = probe.Confidence
	r.Reason = probe.Reason
	r.CurrentN = probe.CurrentN
	r.RecommendedN = probe.RecommendedN
	r.Tradeoffs = probe.Tradeoffs
	return nil
}

// ComparisonReport is the full comparison report for one candidate config
// (Rule 25 includes TraceLinks).
type ComparisonReport struct {
	BaselineConfigID  string             `json:"baseline_config_id"`
	CandidateConfigID string             `json:"candidate_config_id"`
	Metrics           []MetricComparison `json:"metrics"`
	Recommendation    Recommendation     `json:"recommendation"`
	TraceLinks        []string           `json:"trace_links"`
}

// ClassifyDirection classifies the direction of a delta relative to the
// metric's optimization direction (Rule 22). A delta within eps of zero is
// NoChange.
func ClassifyDirection(delta float64, direction sporecore.OptimizationDirection, eps float64) ComparisonDirection {
	if math.Abs(delta) <= eps {
		return DirectionNoChange
	}
	switch direction {
	case sporecore.OptimizationMaximize:
		if delta > 0.0 {
			return DirectionBetter
		}
		return DirectionWorse
	default: // Minimize
		if delta < 0.0 {
			return DirectionBetter
		}
		return DirectionWorse
	}
}

// SignificanceAlpha is the significance threshold for "improved"/"regressed"
// (Rule 23).
const SignificanceAlpha = 0.05

// DeriveRecommendation derives a Recommendation from the per-metric comparisons
// and the primary metric (Rules 23-24).
//
//   - Primary improves with p < 0.05            -> Adopt.
//   - Primary clearly worse (Worse, p < 0.05)   -> Reject.
//   - Primary inconclusive (p >= 0.05)          -> NeedsMoreRuns.
//   - Otherwise mixed across metrics            -> Ambiguous.
func DeriveRecommendation(candidateConfigID string, comparisons []MetricComparison, primary metricmap.EvalMetric) Recommendation {
	primaryName := primary.Name()
	var primaryCmp *MetricComparison
	for i := range comparisons {
		if comparisons[i].MetricName == primaryName {
			primaryCmp = &comparisons[i]
			break
		}
	}
	if primaryCmp == nil {
		return Recommendation{
			Kind:      RecAmbiguous,
			Tradeoffs: []string{fmt.Sprintf("primary metric %s not measured", primaryName)},
		}
	}

	significant := primaryCmp.PValue < SignificanceAlpha

	if significant && primaryCmp.Direction == DirectionBetter {
		return Recommendation{Kind: RecAdopt, ConfigID: candidateConfigID, Confidence: 1.0 - primaryCmp.PValue}
	}
	if significant && primaryCmp.Direction == DirectionWorse {
		return Recommendation{
			Kind:   RecReject,
			Reason: fmt.Sprintf("primary metric %s regressed (delta=%.4f, p=%.4f)", primaryName, primaryCmp.Delta, primaryCmp.PValue),
		}
	}

	// Inconclusive on the primary metric. If other metrics disagree in
	// direction, flag the tradeoffs; otherwise ask for more runs.
	var mixed []string
	anyBetter, anyWorse := false, false
	for i := range comparisons {
		c := &comparisons[i]
		if c.Direction != DirectionNoChange {
			mixed = append(mixed, fmt.Sprintf("%s: %s (p=%.3f)", c.MetricName, c.Direction, c.PValue))
		}
		if c.Direction == DirectionBetter {
			anyBetter = true
		}
		if c.Direction == DirectionWorse {
			anyWorse = true
		}
	}
	if anyBetter && anyWorse {
		if mixed == nil {
			mixed = []string{}
		}
		return Recommendation{Kind: RecAmbiguous, Tradeoffs: mixed}
	}
	return Recommendation{
		Kind:         RecNeedsMoreRuns,
		CurrentN:     primaryCmp.Candidate.N,
		RecommendedN: RecommendedN(primaryCmp),
	}
}

// RecommendedN is the power-based estimate of the runs needed to detect the
// observed effect at alpha=0.05, power ~= 0.8 (Rule 24): n ~= 16 * sigma^2 /
// delta^2. Pooled variance from baseline and candidate; clamped to a floor
// above the current n.
func RecommendedN(cmp *MetricComparison) uint32 {
	pooledVar := (cmp.Baseline.Stddev*cmp.Baseline.Stddev + cmp.Candidate.Stddev*cmp.Candidate.Stddev) / 2.0
	delta := math.Abs(cmp.Delta)
	current := cmp.Candidate.N
	if current < 1 {
		current = 1
	}
	if delta <= 2.220446049250313e-16 || pooledVar <= 0.0 {
		twice := current * 2
		if twice < current+1 {
			twice = current + 1
		}
		return twice
	}
	est := uint32(math.Ceil(16.0 * pooledVar / (delta * delta)))
	if est < current+1 {
		est = current + 1
	}
	return est
}

File: python/packages/spore_eval/src/spore_eval/eval/report.py
"""Comparison + recommendation types and derivation (Rules 19-25).

Mirrors ``rust/crates/spore-eval/src/report.rs``.
"""

from __future__ import annotations

import math
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field
from spore_core.harness import OptimizationDirection

from .metric_map import EvalMetric, metric_name
from .stats import ConfidenceInterval, MetricStats


class _Model(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)


# ============================================================================
# ComparisonDirection
# ============================================================================

ComparisonDirection = Literal["better", "worse", "no_change"]


# ============================================================================
# MetricComparison
# ============================================================================


class MetricComparison(_Model):
    """One metric's baseline-vs-candidate comparison (Rules 19-22)."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True, arbitrary_types_allowed=True)

    metric_name: str
    baseline: MetricStats
    candidate: MetricStats
    delta: float
    p_value: float
    ci: ConfidenceInterval | None = None
    direction: ComparisonDirection


# ============================================================================
# Recommendation — tagged union on ``kind`` (Rules 23-24)
# ============================================================================


class RecommendationAdopt(_Model):
    kind: Literal["adopt"] = "adopt"
    config_id: str
    confidence: float


class RecommendationReject(_Model):
    kind: Literal["reject"] = "reject"
    reason: str


class RecommendationNeedsMoreRuns(_Model):
    kind: Literal["needs_more_runs"] = "needs_more_runs"
    current_n: int
    recommended_n: int


class RecommendationAmbiguous(_Model):
    kind: Literal["ambiguous"] = "ambiguous"
    tradeoffs: list[str] = Field(default_factory=list)


Recommendation = Annotated[
    RecommendationAdopt
    | RecommendationReject
    | RecommendationNeedsMoreRuns
    | RecommendationAmbiguous,
    Field(discriminator="kind"),
]


# ============================================================================
# ComparisonReport
# ============================================================================


class ComparisonReport(_Model):
    """The full comparison report for one candidate config (Rule 25 includes
    ``trace_links``)."""

    baseline_config_id: str
    candidate_config_id: str
    model_config = ConfigDict(extra="forbid", populate_by_name=True, arbitrary_types_allowed=True)

    metrics: list[MetricComparison] = Field(default_factory=list)
    recommendation: Recommendation
    trace_links: list[str] = Field(default_factory=list)


# ============================================================================
# Derivation
# ============================================================================


def classify_direction(
    delta: float, direction: OptimizationDirection, eps: float = 1e-9
) -> ComparisonDirection:
    """Classify the direction of a delta relative to the metric's optimization
    direction (Rule 22). A delta within ``eps`` of zero is ``no_change``."""
    if abs(delta) <= eps:
        return "no_change"
    if direction == "maximize":
        return "better" if delta > 0.0 else "worse"
    return "better" if delta < 0.0 else "worse"


SIGNIFICANCE_ALPHA = 0.05
"""The significance threshold for "improved" / "regressed" (Rule 23)."""


def derive_recommendation(
    candidate_config_id: str,
    comparisons: list[MetricComparison],
    primary: EvalMetric,
) -> Recommendation:
    """Derive a :data:`Recommendation` from the per-metric comparisons and the
    primary metric (Rules 23-24).

    * Primary improves with p < 0.05            → ``adopt``.
    * Primary clearly worse (worse, p < 0.05)   → ``reject``.
    * Primary inconclusive (p >= 0.05)          → ``needs_more_runs``.
    * Otherwise mixed across metrics            → ``ambiguous``."""
    primary_name = metric_name(primary)
    primary_cmp = next((c for c in comparisons if c.metric_name == primary_name), None)
    if primary_cmp is None:
        return RecommendationAmbiguous(tradeoffs=[f"primary metric {primary_name} not measured"])

    significant = primary_cmp.p_value < SIGNIFICANCE_ALPHA

    if significant and primary_cmp.direction == "better":
        return RecommendationAdopt(
            config_id=candidate_config_id,
            confidence=1.0 - primary_cmp.p_value,
        )
    if significant and primary_cmp.direction == "worse":
        return RecommendationReject(
            reason=(
                f"primary metric {primary_name} regressed "
                f"(delta={primary_cmp.delta:.4f}, p={primary_cmp.p_value:.4f})"
            )
        )

    # Inconclusive on the primary metric.
    mixed = [
        f"{c.metric_name}: {c.direction} (p={c.p_value:.3f})"
        for c in comparisons
        if c.direction != "no_change"
    ]
    any_better = any(c.direction == "better" for c in comparisons)
    any_worse = any(c.direction == "worse" for c in comparisons)
    if any_better and any_worse:
        return RecommendationAmbiguous(tradeoffs=mixed)
    return RecommendationNeedsMoreRuns(
        current_n=primary_cmp.candidate.n,
        recommended_n=recommended_n(primary_cmp),
    )


def recommended_n(cmp: MetricComparison) -> int:
    """Power-based estimate of the runs needed to detect the observed effect at
    alpha=0.05, power~=0.8 (Rule 24): ``n ~= 16 sigma^2 / delta^2``. Pooled
    variance from baseline and candidate; clamped above the current n."""
    pooled_var = (cmp.baseline.stddev**2 + cmp.candidate.stddev**2) / 2.0
    delta = abs(cmp.delta)
    current = max(cmp.candidate.n, 1)
    if delta <= 2.220446049250313e-16 or pooled_var <= 0.0:
        return max(current * 2, current + 1)
    est = math.ceil(16.0 * pooled_var / (delta * delta))
    return max(est, current + 1)
